crates/net/network/src/manager.rs

```diff
@@ -596,6 +597,34 @@ where
     }
 }
 
+impl<C> NetworkManager<C>
+where
+    C: BlockReader + Unpin,
+{
+    /// Drives the [NetworkManager] future until a [GracefulShutdown] signal is received.
+    ///
+    /// This also run the given function `shutdown_hook` afterwards.
+    pub async fn run_until_graceful_shutdown(
+        self,
+        shutdown: GracefulShutdown,
+        shutdown_hook: impl FnOnce(&mut Self),
```

Comment:
we might wanna change this to a future, but can do this separately as good first issue